src/test/java/fi/aalto/cs/apluscourses/presentation/base/SelectableNodeViewModelTest.java

```diff
@@ -94,4 +96,13 @@ public void testApplyFilterReturnsEmpty() throws InterruptedException {
     assertFalse(result.isPresent());
     assertTrue(node.isVisible());
   }
+
+  @Test
+  public void testApplyFilterHidesEmptyNodesThatWantToGetHidden() throws InterruptedException {
+    Filter filter = mock(Filter.class, new Returns(Optional.of(false)));
+
+    Optional<Boolean> result = emptyNode.applyFilter(filter);
+    assertTrue(result.isPresent());
+    assertFalse(emptyNode.isVisible());
+  }
```

Comment:
This tests more `TestNodeViewModel` than anything else. If we want to test `ExerciseGroupViewModel`'s special case that should happen in `ExerciseGroupViewModelTest`. (You could test `setVisibilityByFilterResult` separately.)